python/ray/data/_internal/plan.py

```diff
@@ -369,19 +369,7 @@ def deep_copy(self) -> "ExecutionPlan":
 
     def initial_num_blocks(self) -> int:
         """Get the estimated number of blocks after applying all plan stages."""
-        if self.has_computed_output():
-            return self._snapshot_blocks.initial_num_blocks()
-        for stage in self._stages_after_snapshot[::-1]:
-            if stage.num_blocks is not None:
-                return stage.num_blocks
-        if self._snapshot_blocks is not None:
-            return self._snapshot_blocks.initial_num_blocks()
-        for stage in self._stages_before_snapshot[::-1]:
-            if stage.num_blocks is not None:
-                return stage.num_blocks
-        if self._in_blocks is not None:
-            return self._in_blocks.initial_num_blocks()
-        return None
+        return self._logical_plan.dag.num_outputs()
```

Comment:
is this still useful? I'm wondering if we should only support `num_blocks` for materialized datasets. Because otherwise the value won't be accurate either.

Reply:
The main usages i could find were:
- `Dataset.num_blocks()` --> since this is public API, i think we should still keep the `initial_num_blocks()` behavior unchanged. I see there is one usage outside Ray Data, which is in Ray Train: https://github.com/ray-project/ray/blob/master/python/ray/train/gbdt_trainer.py#L256 
- `Dataset._tab_repr_()` --> we could probably omit this if we wanted to drop it.

Reply:
from user's perspective, if they have not executed a Dataset yet, would they always see something like `Dataset(..., num_blocks=?, ...)`? I was under the impression that we wanted to use the best known estimate (which we are getting from `ds._plan.initial_num_blocks()`) so that we would avoid `num_blocks=?` when possible -- even though this may not be the most accurate number

Reply:
as we migrated to streaming execution, I think it's no longer necessary to expose `num_blocks` to users. Because it's inaccurate anyway. and users are not likely to need this info. For the Ray Train usage, instead we can have an API to set the number of consumers and do repartition internally. So I think it makes sense to deprecate this API. 
We can merge this PR first and discuss more on this issue https://github.com/ray-project/ray/issues/42184.